### ISSUE TITLE
Kyber-AVX2 doesn't work on OS X, Windows

### ISSUE THREAD

**Issue by thomwiggers** [PQClean/PQClean#251]
The Kyber ASM code doesn't work on OS X (macros don't seem to be supported) nor on Windows (Microsoft compiler has other syntax).

**Comment by thomwiggers** [comment]
The MacOS tests have not been running because the Travis Mac cpus are quite old and don't have AVX2.

**Comment by thomwiggers** [comment]
This can probably be solved in the same way as Dilithium. The actual problem is that C symbols are prefixed with an underscore on MacOS

**Comment by thomwiggers** [comment]
OS X compatibility changes required:

* Add underscore equivalents of symbols in assembly functions or do it dynamically with a macro (the C ABI starts with an underscore on OS X) https://github.com/PQClean/PQClean/blob/master/crypto_sign/dilithium2/avx2/ntt.S#L44-L45* Make sure any symbols referred to in ASM code that are defined in C files (constants?) are wrapped in a macro like https://github.com/PQClean/PQClean/blob/master/crypto_sign/dilithium2/avx2/ntt.S#L47-L49

**Comment by mouse07410** [comment]
Since this issue appears to be quite old - any chance to get the fix sooner...?

```
$ make SCHEME=Kyber1024 IMPLEMENTATION=avx2
cd ../crypto_kem/Kyber1024/avx2 && /Applications/Xcode.app/Contents/Developer/usr/bin/make
clang -mavx2 -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o kem.o kem.c
clang -mavx2 -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o poly.o poly.c
clang -mavx2 -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o polyvec.o polyvec.c
yasm -m amd64 -f macho64 -p gas  -o fq.o fq.s
yasm -m amd64 -f macho64 -p gas  -o shuffle.o shuffle.s
clang -mavx2 -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o cbd.o cbd.c
yasm -m amd64 -f macho64 -p gas  -o ntt.o ntt.s
yasm -m amd64 -f macho64 -p gas  -o invntt.o invntt.s
yasm -m amd64 -f macho64 -p gas  -c -o basemul.o basemul.S
yasm: warning: unrecognized option `-c'
clang -mavx2 -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o consts.o consts.c
clang -mavx2 -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o verify.o verify.c
clang -mavx2 -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o indcpa.o indcpa.c
clang -mavx2 -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o rejsample.o rejsample.c
clang -mavx2 -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o fips202x4.o fips202x4.c
clang -mavx2 -mbmi2 -mpopcnt -O3 -Wall -Wextra -Wpedantic -Werror -Wmissing-prototypes -Wredundant-decls -std=c99 -I../../../common  -c -o symmetric-fips202.o symmetric-fips202.c
ar -r libkyber1024_avx2.a kem.o poly.o polyvec.o fq.o shuffle.o cbd.o ntt.o invntt.o basemul.o consts.o verify.o indcpa.o rejsample.o fips202x4.o symmetric-fips202.o ../../../common/keccak4x/KeccakP-1600-times4-SIMD256.o
ar: creating archive libkyber1024_avx2.a
mkdir -p ../bin
clang -O3 -Wall -Wextra -Wpedantic -Wvla -Werror -std=c99 -Wundef -Wshadow -Wcast-align -Wpointer-arith -Wmissing-prototypes -fstrict-aliasing -fno-common -pipe -I../common  -DNTESTS=1 -DPQCLEAN_NAMESPACE=PQCLEAN_KYBER1024_AVX2 -I../crypto_kem/Kyber1024/avx2 crypto_kem/functest.c ../common/aes.c ../common/sha2.c ../common/fips202.c ../common/nistseedexpander.c ../common/sp800-185.c ../common/randombytes.c -o ../bin/functest_Kyber1024_avx2 -L../crypto_kem/Kyber1024/avx2 -lKyber1024_avx2
Undefined symbols for architecture x86_64:
  "_PQCLEAN_KYBER1024_AVX2_basemul_acc_avx", referenced from:
      _PQCLEAN_KYBER1024_AVX2_polyvec_pointwise_acc in libKyber1024_avx2.a(polyvec.o)
  "_PQCLEAN_KYBER1024_AVX2_basemul_avx", referenced from:
      _PQCLEAN_KYBER1024_AVX2_poly_basemul in libKyber1024_avx2.a(poly.o)
  "_PQCLEAN_KYBER1024_AVX2_csubq_avx", referenced from:
      _PQCLEAN_KYBER1024_AVX2_poly_compress in libKyber1024_avx2.a(poly.o)
      _PQCLEAN_KYBER1024_AVX2_poly_csubq in libKyber1024_avx2.a(poly.o)
  "_PQCLEAN_KYBER1024_AVX2_frommont_avx", referenced from:
      _PQCLEAN_KYBER1024_AVX2_poly_frommont in libKyber1024_avx2.a(poly.o)
  "_PQCLEAN_KYBER1024_AVX2_invntt_level6_avx", referenced from:
      _PQCLEAN_KYBER1024_AVX2_poly_invntt in libKyber1024_avx2.a(poly.o)
  "_PQCLEAN_KYBER1024_AVX2_invntt_levels0t5_avx", referenced from:
      _PQCLEAN_KYBER1024_AVX2_poly_invntt in libKyber1024_avx2.a(poly.o)
  "_PQCLEAN_KYBER1024_AVX2_ntt_level0_avx", referenced from:
      _PQCLEAN_KYBER1024_AVX2_poly_ntt in libKyber1024_avx2.a(poly.o)
  "_PQCLEAN_KYBER1024_AVX2_ntt_levels1t6_avx", referenced from:
      _PQCLEAN_KYBER1024_AVX2_poly_ntt in libKyber1024_avx2.a(poly.o)
  "_PQCLEAN_KYBER1024_AVX2_nttfrombytes_avx", referenced from:
      _PQCLEAN_KYBER1024_AVX2_poly_frombytes in libKyber1024_avx2.a(poly.o)
  "_PQCLEAN_KYBER1024_AVX2_ntttobytes_avx", referenced from:
      _PQCLEAN_KYBER1024_AVX2_poly_tobytes in libKyber1024_avx2.a(poly.o)
  "_PQCLEAN_KYBER1024_AVX2_nttunpack_avx", referenced from:
      _PQCLEAN_KYBER1024_AVX2_poly_nttunpack in libKyber1024_avx2.a(poly.o)
  "_PQCLEAN_KYBER1024_AVX2_reduce_avx", referenced from:
      _PQCLEAN_KYBER1024_AVX2_poly_reduce in libKyber1024_avx2.a(poly.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [../bin/functest_Kyber1024_avx2] Error 1
```

**Comment by mouse07410** [comment]
I bit the bullet and tried a simple workaround. It worked, but the diff is about 960 lines! Of this kind:
```diff
diff --git a/crypto_kem/kyber768/avx2/shuffle.s b/crypto_kem/kyber768/avx2/shuffle.s
index 6a8edecb..bcfdcebe 100644
--- a/crypto_kem/kyber768/avx2/shuffle.s
+++ b/crypto_kem/kyber768/avx2/shuffle.s
@@ -1,11 +1,11 @@
 .include "fq.inc"
 .include "shuffle.inc"
 
-.global PQCLEAN_KYBER768_AVX2_nttunpack_avx
-PQCLEAN_KYBER768_AVX2_nttunpack_avx:
+.global _PQCLEAN_KYBER768_AVX2_nttunpack_avx
+_PQCLEAN_KYBER768_AVX2_nttunpack_avx:
 #consts
-vmovdqa                PQCLEAN_KYBER768_AVX2_16xq(%rip),%ymm0
-vmovdqa                PQCLEAN_KYBER768_AVX2_16xv(%rip),%ymm1
+vmovdqa                _PQCLEAN_KYBER768_AVX2_16xq(%rip),%ymm0
+vmovdqa                _PQCLEAN_KYBER768_AVX2_16xv(%rip),%ymm1
 
 #load
 vmovdqa                (%rdi),%ymm4
```

Complete diff to get AVX2 version of Kyber run on macOS: [kyber-avx2.diff.txt](https://github.com/PQClean/PQClean/files/5258098/kyber-avx2.diff.txt)

It looks like the correct fix would be in the `crypto_kem/kyber1024/avx2/Makefile` or such?

**Comment by mouse07410** [comment]
Oh, it's probably time to close this one.

**Comment by thomwiggers** [comment]
This will be closed when #307 is merged

**Comment by baentsch** [comment]
Just came across this issue -- seems resolved by #349 -> close?
